README.md
# Overview

This is an [LZ4](http://lz4.org)-frame compression library for Python v3.2+ (and 2.7+), bound to Yann Collet's [LZ4 C implementation](https://github.com/lz4/lz4).


# Installing / packaging
```shell
# To get from PyPI
pip3 install py-lz4framed

# To only build extension modules inline (e.g. in repository)
python3 setup.py build_ext -i

# To build & install globally
python3 setup.py install
```
**Notes**

- The above as well as all other python3-using commands should also run with v2.7+
- This module is also available via [Anaconda (conda-forge)](https://anaconda.org/conda-forge/py-lz4framed) (with binaries for Linux, OSX and Windows)
- PyPI releases are signed with the [Iotic Labs Software release signing key](https://developer.iotic-labs.com/iotic-labs.com.asc)


# Usage
Single-function operation:
```python
import lz4framed

compressed = lz4framed.compress(b'binary data')

uncompressed = lz4framed.decompress(compressed)
```
To iteratively compress (to a file or e.g. BytesIO instance):
```python
with open('myFile', 'wb') as f:
    # Context automatically finalises frame on completion, unless an exception occurs
    with Compressor(f) as c:
        try:
            while (...):
               c.update(moreData)
        except Lz4FramedNoDataError:
            pass
```
To decompress from a file-like object:
```python
with open('myFile', 'rb') as f:
    try:
        for chunk in Decompressor(f):
           decoded.append(chunk)
    except Lz4FramedNoDataError:
        # Compress frame data incomplete - error case
        ...
```
See also [lz4framed/\_\_main\_\_.py](lz4framed/__main__.py) for example usage.

# Documentation
```python
import lz4framed
print(lz4framed.__version__, lz4framed.LZ4_VERSION, lz4framed.LZ4F_VERSION)
help(lz4framed)
```

# Command-line utility
```shell
python3 -mlz4framed
USAGE: lz4framed (compress|decompress) (INFILE|-) [OUTFILE]

(De)compresses an lz4 frame. Input is read from INFILE unless set to '-', in
which case stdin is used. If OUTFILE is not specified, output goes to stdout.
```


# Tests

## Static
This library has been checked using [flake8](https://pypi.python.org/pypi/flake8) and [pylint](http://www.pylint.org), using a modified configuration - see _pylint.rc_ and _flake8.cfg_.

## Unit
```shell
python3 -m unittest discover -v .
```

# Why?
The only existing lz4-frame interoperable implementation I was aware of at the time of writing ([lz4tools](https://github.com/darkdragn/lz4tools)) had the following limitations:

- Incomplete implementation in terms of e.g. reference & memory leaks on failure
- Lack of unit tests
- Not thread safe
- Does not release GIL during low level (de)compression operations
- Did not address the requirements for an external project


CHANGELOG
0.14.0
- Updated lz4 to v1.9.2 (includes fix for CVE-2019-17543)

0.13.0
- Updated lz4 to v1.8.3 (data corruption fix for v1.8.2)

0.12.0
- Updated lz4 to v1.8.2
- Confirmed compatibility with Python 3.7

0.11.0
- Updated lz4 to v1.8.1 (faster/stronger ultra modes, levels 10+)
- Minor documentation clarification about compression levels.

0.10.0
- Updated lz4 to v1.8.0
- Support for fast acceleration via negative compression levels
- Support for block checksums

0.9.7
- Allow for any bytes-like objects as input (e.g. memoryview)

0.9.6
- Windows build compatibility

0.9.5
- Potential reference leak fix (on LZ4FError being raised)

0.9.4
- Updated lz4 to v1.7.5 (Note: Available compression levels have changed)
- Use lz4 constants for available compression levels
- Use LZ4_VERSION_STRING to propagate underlying lz4 version
- Update tests to reflect new compression levels

0.9.3
- Allow for lack of stdin/stdout/stdout buffer access (#3, https://github.com/windreamer)

0.9.2
- Fix index in argv (Jan Špaček)
- License information update

0.9.1
- Fixup Python 2/3 compat layer for nose (Mathias Laurin)

0.9.0
- Initial public release


test.py
# Copyright (c) 2016 Iotic Labs Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/Iotic-Labs/py-lz4framed/blob/master/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""Note: These tests are not meant to verify all of lz4's behaviour, only the Python functionality"""

from sys import version_info
from unittest import TestCase
from contextlib import contextmanager
from io import BytesIO, SEEK_END

from lz4framed import (LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB, LZ4F_BLOCKSIZE_MAX1MB,
                       LZ4F_BLOCKSIZE_MAX4MB,
                       LZ4F_COMPRESSION_MAX,
                       LZ4F_ERROR_GENERIC, LZ4F_ERROR_contentChecksum_invalid, LZ4F_ERROR_frameType_unknown,
                       LZ4F_ERROR_srcPtr_wrong, Lz4FramedError, Lz4FramedNoDataError,
                       compress, decompress,
                       create_compression_context, compress_begin, compress_update, compress_end,
                       create_decompression_context, get_frame_info, decompress_update,
                       get_block_size,
                       Compressor, Decompressor)

PY2 = version_info[0] < 3

SHORT_INPUT = b'abcdefghijklmnopqrstuvwxyz0123456789'
LONG_INPUT = SHORT_INPUT * (10**5)
LEVEL_ACCELERATED_MAX = -10


class TestHelperMixin(object):

    def setUp(self):
        super(TestHelperMixin, self).setUp()
        if PY2:
            # avoid deprecation warning
            self.assertRaisesRegex = self.assertRaisesRegexp  # pylint: disable=invalid-name

    def check_compress_short(self, *args, **kwargs):
        self.assertEqual(SHORT_INPUT, decompress(compress(SHORT_INPUT, *args, **kwargs)))

    def check_compress_long(self, *args, **kwargs):
        self.assertEqual(LONG_INPUT, decompress(compress(LONG_INPUT, *args, **kwargs)))

    @contextmanager
    def assertRaisesLz4FramedError(self, code):  # pylint: disable=invalid-name
        try:
            yield
        except Lz4FramedError as ex:
            self.assertEqual(ex.args[1], code, 'Lz4FramedError code mismatch: [%d]: %s' % (ex.args[1], ex.args[0]))
        else:
            self.fail('Lz4FramedError not raised')


class TestCompress(TestHelperMixin, TestCase):

    def test_compress_minimal(self):
        with self.assertRaises(TypeError):
            compress()
        with self.assertRaises(Lz4FramedNoDataError):
            compress(b'')
        self.check_compress_short()

    def test_compress_block_size(self):
        with self.assertRaises(TypeError):
            compress(SHORT_INPUT, block_size_id='1')
        with self.assertRaises(ValueError):
            compress(SHORT_INPUT, block_size_id=-1)
        for block_size in (LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB,
                           LZ4F_BLOCKSIZE_MAX1MB, LZ4F_BLOCKSIZE_MAX4MB):
            self.check_compress_short(block_size_id=block_size)
            self.check_compress_long(block_size_id=block_size)

    def test_compress_linked_mode(self):
        with self.assertRaises(TypeError):
            compress(SHORT_INPUT, block_mode_linked=None)
        self.check_compress_short(block_mode_linked=True)
        self.check_compress_short(block_mode_linked=False)

    def test_compress_checksum(self):
        with self.assertRaises(TypeError):
            compress(SHORT_INPUT, checksum=None)
        self.check_compress_short(checksum=True)
        self.check_compress_short(checksum=False)
        for data in (SHORT_INPUT, LONG_INPUT):
            with self.assertRaisesLz4FramedError(LZ4F_ERROR_contentChecksum_invalid):
                # invalid checksum
                decompress(compress(data, checksum=True)[:-1] + b'0')

    def test_compress_block_checksum(self):
        with self.assertRaises(TypeError):
            compress(SHORT_INPUT, block_checksum=None)
        self.check_compress_short(block_checksum=True)
        self.check_compress_short(block_checksum=False)

    def test_compress_level(self):
        with self.assertRaises(TypeError):
            compress(SHORT_INPUT, level='1')
        # negative values designate accelerattion
        for level in range(LEVEL_ACCELERATED_MAX, LZ4F_COMPRESSION_MAX + 1):
            self.check_compress_short(level=level)
        # large input, fast & hc levels (levels > 10 (v1.7.5) are significantly slower)
        self.check_compress_long(level=0)
        self.check_compress_long(level=10)

    def test_compress_memoryview(self):
        view = memoryview(LONG_INPUT)
        self.assertEqual(view, decompress(compress(view)))


class TestDecompress(TestHelperMixin, TestCase):

    def test_decompress_minimal(self):
        with self.assertRaises(TypeError):
            decompress()
        with self.assertRaises(Lz4FramedNoDataError):
            decompress(b'')
        self.check_compress_short()

    def test_decompress_buffer_size(self):
        out = compress(SHORT_INPUT)
        with self.assertRaises(TypeError):
            decompress(out, buffer_size='1')
        with self.assertRaises(ValueError):
            decompress(out, buffer_size=0)
        out = compress(LONG_INPUT)
        for buffer_size in range(1, 1025, 128):
            self.assertEqual(LONG_INPUT, decompress(out, buffer_size=buffer_size))

    def test_decompress_invalid_input(self):
        with self.assertRaisesLz4FramedError(LZ4F_ERROR_frameType_unknown):
            decompress(b'invalidheader')
        with self.assertRaisesRegex(ValueError, 'frame incomplete'):
            decompress(compress(SHORT_INPUT)[:-5])
        # incomplete data (length not specified in header)
        with BytesIO() as out:
            with Compressor(out) as compressor:
                compressor.update(SHORT_INPUT)
            output = out.getvalue()
            with self.assertRaisesRegex(ValueError, 'frame incomplete'):
                decompress(output[:-20])

    def test_decompress_memoryview(self):
        view = memoryview(compress(LONG_INPUT))
        self.assertEqual(LONG_INPUT, decompress(view))


class TestLowLevelFunctions(TestHelperMixin, TestCase):

    def test_get_block_size(self):
        with self.assertRaises(TypeError):
            get_block_size('1')
        with self.assertRaises(ValueError):
            get_block_size(1)
        self.assertEqual(get_block_size(), get_block_size(LZ4F_BLOCKSIZE_DEFAULT))
        for size in (LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB, LZ4F_BLOCKSIZE_MAX1MB, LZ4F_BLOCKSIZE_MAX4MB):
            self.assertEqual(get_block_size(size), 1 << (8 + (2 * size)))

    def test_create_contexts(self):
        for func in (create_compression_context, create_decompression_context):
            self.assertIsNotNone(func())

    def test_get_frame_info(self):
        with self.assertRaises(TypeError):
            get_frame_info()
        with self.assertRaises(ValueError):
            get_frame_info(create_compression_context())

        ctx = create_decompression_context()
        with self.assertRaisesLz4FramedError(LZ4F_ERROR_srcPtr_wrong):
            get_frame_info(ctx)
        # compress with non-default arguments, check info structure
        args = {'checksum': True,
                'block_size_id': LZ4F_BLOCKSIZE_MAX256KB,
                'block_mode_linked': False}
        # Using long input since lz4 adjusts block size is input smaller than one block
        decompress_update(ctx, compress(LONG_INPUT, **args)[:15])
        info = get_frame_info(ctx)
        self.assertTrue(info.pop('input_hint', 0) > 0)
        args['length'] = len(LONG_INPUT)
        self.assertEqual(info, args)

    def __compress_begin(self, **kwargs):
        ctx = create_compression_context()
        header = compress_begin(ctx, **kwargs)
        self.assertTrue(7 <= len(header) <= 15)
        return ctx, header

    def test_compress_begin(self):
        with self.assertRaises(TypeError):
            compress_begin()
        with self.assertRaises(ValueError):
            compress_begin(create_decompression_context())

    def test_compress_begin_block_size(self):
        with self.assertRaises(TypeError):
            self.__compress_begin(block_size_id='1')
        with self.assertRaises(ValueError):
            self.__compress_begin(block_size_id=-1)
        for size in (LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB, LZ4F_BLOCKSIZE_MAX1MB,
                     LZ4F_BLOCKSIZE_MAX4MB):
            self.__compress_begin(block_size_id=size)

    def test_compress_begin_linked_mode(self):
        with self.assertRaises(TypeError):
            self.__compress_begin(block_mode_linked=None)
        self.__compress_begin(block_mode_linked=True)
        self.__compress_begin(block_mode_linked=False)

    def test_compress_begin_checksum(self):
        with self.assertRaises(TypeError):
            self.__compress_begin(checksum=None)
        self.__compress_begin(checksum=True)
        self.__compress_begin(checksum=False)

    def test_compress_begin_level(self):
        with self.assertRaises(TypeError):
            self.__compress_begin(level='1')
        for level in range(LEVEL_ACCELERATED_MAX, LZ4F_COMPRESSION_MAX + 1):
            self.__compress_begin(level=level)

    def test_compress_update_invalid(self):
        with self.assertRaises(TypeError):
            compress_update()
        with self.assertRaises(TypeError):
            compress_update(1)
        # invalid context
        with self.assertRaises(ValueError):
            compress_update(create_decompression_context(), b' ')
        # data before compress_begin called
        with self.assertRaisesLz4FramedError(LZ4F_ERROR_GENERIC):
            compress_update(create_compression_context(), b' ')

        ctx, _ = self.__compress_begin()
        # invalid data
        with self.assertRaises(TypeError):
            compress_update(ctx, 1)
        # empty data
        with self.assertRaises(Lz4FramedNoDataError):
            compress_update(ctx, b'')

    def test_compress_end(self):
        with self.assertRaises(TypeError):
            compress_end()
        with self.assertRaises(ValueError):
            compress_end(create_decompression_context())

        ctx, header = self.__compress_begin()
        self.assertEqual(b'', decompress(header + compress_end(ctx)))

        ctx, header = self.__compress_begin()
        data = compress_update(ctx, SHORT_INPUT)
        self.assertEqual(decompress(header + data + compress_end(ctx)), SHORT_INPUT)

    def __compress_with_data_and_args(self, data, **kwargs):
        ctx, header = self.__compress_begin(**kwargs)
        in_raw = BytesIO(data)
        out = BytesIO(header)
        out.seek(0, SEEK_END)
        try:
            while True:
                out.write(compress_update(ctx, in_raw.read(1024)))
        except Lz4FramedNoDataError:
            pass
        out.write(compress_end(ctx))
        self.assertEqual(decompress(out.getvalue()), data)

    def test_compress(self):
        func = self.__compress_with_data_and_args

        for size in (LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB, LZ4F_BLOCKSIZE_MAX1MB,
                     LZ4F_BLOCKSIZE_MAX4MB):
            func(LONG_INPUT, block_size_id=size)

        for arg in ('block_mode_linked', 'checksum'):
            for value in (False, True):
                func(LONG_INPUT, **{arg: value})

        for level in range(LEVEL_ACCELERATED_MAX, LZ4F_COMPRESSION_MAX + 1):
            func(SHORT_INPUT, level=level)

        func(memoryview(LONG_INPUT))

    def test_decompress_update_invalid(self):
        with self.assertRaises(TypeError):
            decompress_update()
        with self.assertRaises(TypeError):
            decompress_update(1)
        # invalid context
        with self.assertRaises(ValueError):
            decompress_update(create_compression_context(), b' ')

        ctx = create_decompression_context()

        with self.assertRaises(TypeError):
            decompress_update(ctx, b' ', chunk_len='1')
        with self.assertRaises(ValueError):
            decompress_update(ctx, b' ', chunk_len=0)

        in_raw = compress(LONG_INPUT, checksum=True)

        ret = decompress_update(ctx, in_raw[:512], chunk_len=2)
        # input_hint
        self.assertTrue(ret.pop() > 0)
        # chunk length
        self.assertTrue(len(ret) > 0)
        self.assertTrue(all(1 <= len(chunk) <= 2 for chunk in ret))

        # invalid input (from start of frame)
        with self.assertRaisesLz4FramedError(LZ4F_ERROR_GENERIC):
            decompress_update(ctx, in_raw)

        # checksum invalid
        in_raw = in_raw[:-4] + b'1234'
        ctx = create_decompression_context()
        with self.assertRaisesLz4FramedError(LZ4F_ERROR_contentChecksum_invalid):
            decompress_update(ctx, in_raw)

    def test_decompress_update_memoryview(self):  # pylint: disable=invalid-name
        ctx = create_decompression_context()
        data = decompress_update(ctx, memoryview(compress(LONG_INPUT)))
        self.assertEqual(b''.join(data[:-1]), LONG_INPUT)


class TestCompressor(TestHelperMixin, TestCase):
    """Note: Low-level methods supporting Compressor class have been tested in TestLowLevelFunctions"""

    def test_compressor_init(self):
        with self.assertRaisesRegex(AttributeError, 'has no attribute \'write\''):
            Compressor('1')

        # non-callable write attribute
        class Empty(object):
            write = 1
        with self.assertRaises(TypeError):
            Compressor(Empty())

        # cannot use context without fp
        with self.assertRaises(ValueError):
            with Compressor() as _:  # noqa (unused variable)
                pass

    def test_compressor__no_fp(self):
        in_bytes = BytesIO(LONG_INPUT)
        out_bytes = BytesIO()

        compressor = Compressor()
        try:
            while True:
                out_bytes.write(compressor.update(in_bytes.read(1024)))
        # raised by compressor.update() on empty data argument
        except Lz4FramedNoDataError:
            pass
        out_bytes.write(compressor.end())

        self.assertEqual(decompress(out_bytes.getvalue()), LONG_INPUT)

    def test_compressor_fp(self):
        self.__fp_test()

    def __fp_test(self, in_raw=LONG_INPUT, **kwargs):
        in_bytes = BytesIO(in_raw)
        out_bytes = BytesIO()

        with Compressor(out_bytes, **kwargs) as compressor:
            try:
                while True:
                    compressor.update(in_bytes.read(1024))
            # raised by compressor.update() on empty data argument
            except Lz4FramedNoDataError:
                pass
        self.assertEqual(decompress(out_bytes.getvalue()), in_raw)

    def test_compressor_block_size(self):
        for block_size in (LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB,
                           LZ4F_BLOCKSIZE_MAX1MB, LZ4F_BLOCKSIZE_MAX4MB):
            self.__fp_test(block_size_id=block_size)

    def test_compressor_checksum(self):
        self.__fp_test(checksum=False)
        self.__fp_test(checksum=True)

    def test_compressor_autoflush(self):
        self.__fp_test(autoflush=True)
        self.__fp_test(autoflush=False)

    def test_compressor_level(self):
        for level in range(LEVEL_ACCELERATED_MAX, LZ4F_COMPRESSION_MAX + 1):
            self.__fp_test(in_raw=SHORT_INPUT, level=level)
        self.__fp_test(level=0)
        # levels > 10 (v1.7.5) are significantly slower
        self.__fp_test(level=10)


class TestDecompressor(TestHelperMixin, TestCase):

    def test_decompressor_init(self):
        with self.assertRaises(TypeError):
            Decompressor()  # pylint: disable=no-value-for-parameter
        with self.assertRaisesRegex(AttributeError, 'has no attribute \'read\''):
            Decompressor('1')

        # non-callable read attribute
        class Empty(object):
            read = 1
        with self.assertRaises(TypeError):
            Decompressor(Empty())

    def test_decompressor_fp(self):
        # levels > 10 (v1.7.5) are significantly slower
        for level in (LEVEL_ACCELERATED_MAX, 10):
            out_bytes = BytesIO()
            for chunk in Decompressor(BytesIO(compress(LONG_INPUT, level=level))):
                out_bytes.write(chunk)
            self.assertEqual(out_bytes.getvalue(), LONG_INPUT)

        # incomplete frame
        out_bytes.truncate()
        with self.assertRaises(Lz4FramedNoDataError):
            for chunk in Decompressor(BytesIO(compress(LONG_INPUT)[:-32])):
                out_bytes.write(chunk)
        # some data should have been written
        out_bytes.seek(SEEK_END)
        self.assertTrue(out_bytes.tell() > 0)


# def pympler_run(iterations=20):
#     from unittest import main
#     from pympler import tracker
#     from gc import collect

#     tracker = tracker.SummaryTracker()
#     for i in range(iterations):
#         try:
#             main()
#         except SystemExit:
#             pass
#         if i % 2:
#             collect()
#             tracker.print_diff()


# if __name__ == '__main__':
#     pympler_run()


lz4framed/compat.py
# Copyright (c) 2016 Iotic Labs Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/Iotic-Labs/py-lz4framed/blob/master/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Original six.py copyright notice, on which snippets herein are based:
#
# Copyright (c) 2010-2015 Benjamin Peterson
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""Python v2.7 (NOT 2.6) compatibility"""

# pylint: disable=unused-import,invalid-name,wrong-import-order

from sys import stderr, stdout, stdin, version_info

try:
    # pylint: disable=no-name-in-module
    from collections.abc import Iterable
except ImportError:
    from collections import Iterable  # noqa

PY2 = (version_info[0] == 2)

if PY2:
    STDIN_RAW = stdin
    STDOUT_RAW = stdout
    STDERR_RAW = stderr

else:
    STDIN_RAW = getattr(stdin, 'buffer', stdin)
    STDOUT_RAW = getattr(stdout, 'buffer', stdout)
    STDERR_RAW = getattr(stderr, 'buffer', stderr)


lz4framed/__main__.py
# Copyright (c) 2016 Iotic Labs Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/Iotic-Labs/py-lz4framed/blob/master/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""(de)compresses to/from lz4-framed data"""

from __future__ import print_function
from sys import argv, stderr

from .compat import STDIN_RAW, STDOUT_RAW
from . import Compressor, Decompressor, Lz4FramedError, Lz4FramedNoDataError, get_block_size


def __error(*args, **kwargs):
    print(*args, file=stderr, **kwargs)


def do_compress(in_stream, out_stream):
    read = in_stream.read
    read_size = get_block_size()
    try:
        with Compressor(out_stream) as compressor:
            try:
                while True:
                    compressor.update(read(read_size))
            # empty read result supplied to update()
            except Lz4FramedNoDataError:
                pass
            # input stream exception
            except EOFError:
                pass
    except Lz4FramedError as ex:
        __error('Compression error: %s' % ex)
        return 8
    return 0


def do_decompress(in_stream, out_stream):
    write = out_stream.write
    try:
        for chunk in Decompressor(in_stream):
            write(chunk)
    except Lz4FramedError as ex:
        __error('Compression error: %s' % ex)
        return 8
    return 0


__ACTION = frozenset(('compress', 'decompress'))


def main():  # noqa (complexity)
    if not (3 <= len(argv) <= 4 and argv[1] in __ACTION):
        print("""USAGE: lz4framed (compress|decompress) (INFILE|-) [OUTFILE]

(De)compresses an lz4 frame. Input is read from INFILE unless set to '-', in
which case stdin is used. If OUTFILE is not specified, output goes to stdout.""", file=stderr)
        return 1

    compress = (argv[1] == 'compress')
    in_file = out_file = None
    try:
        # input
        if argv[2] == '-':
            in_stream = STDIN_RAW
        else:
            try:
                in_stream = in_file = open(argv[2], 'rb')
            except IOError as ex:
                __error('Failed to open input file for reading: %s' % ex)
                return 2
        # output
        if len(argv) == 3:
            out_stream = STDOUT_RAW
        else:
            try:
                out_stream = out_file = open(argv[3], 'ab')
            except IOError as ex:
                __error('Failed to open output file for appending: %s' % ex)
                return 4

        return (do_compress if compress else do_decompress)(in_stream, out_stream)
    except IOError as ex:
        __error('I/O failure: %s' % ex)
    finally:
        if in_file:
            in_file.close()
        if out_file:
            out_file.close()


if __name__ == "__main__":
    exit(main())


lz4framed/__init__.py
# Copyright (c) 2016 Iotic Labs Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/Iotic-Labs/py-lz4framed/blob/master/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""lz4 frame compression library, bound to lz4 C implementation

Example usage:

# To compress
compressed = lz4framed.compress(b'binary data')

# To decode
uncompressed = lz4framed.decompress(compressed)

To use a file-like objects as input/output, use the provided Compressor & Decompressor
classes instead or manually utilise the context-using low-level methods. All methods are thread safe unless stated.
"""

from threading import Lock

from _lz4framed import (  # noqa (unused import)
    LZ4F_BLOCKSIZE_DEFAULT, LZ4F_BLOCKSIZE_MAX64KB, LZ4F_BLOCKSIZE_MAX256KB, LZ4F_BLOCKSIZE_MAX1MB,
    LZ4F_BLOCKSIZE_MAX4MB,
    LZ4F_COMPRESSION_MIN, LZ4F_COMPRESSION_MIN_HC, LZ4F_COMPRESSION_MAX,
    LZ4F_ERROR_GENERIC, LZ4F_ERROR_maxBlockSize_invalid, LZ4F_ERROR_blockMode_invalid,
    LZ4F_ERROR_contentChecksumFlag_invalid, LZ4F_ERROR_compressionLevel_invalid, LZ4F_ERROR_headerVersion_wrong,
    LZ4F_ERROR_blockChecksum_invalid, LZ4F_ERROR_reservedFlag_set, LZ4F_ERROR_allocation_failed,
    LZ4F_ERROR_srcSize_tooLarge, LZ4F_ERROR_dstMaxSize_tooSmall, LZ4F_ERROR_frameHeader_incomplete,
    LZ4F_ERROR_frameType_unknown, LZ4F_ERROR_frameSize_wrong, LZ4F_ERROR_srcPtr_wrong, LZ4F_ERROR_decompressionFailed,
    LZ4F_ERROR_headerChecksum_invalid, LZ4F_ERROR_contentChecksum_invalid, LZ4F_ERROR_frameDecoding_alreadyStarted,
    LZ4F_VERSION, LZ4_VERSION, __version__,
    Lz4FramedError, Lz4FramedNoDataError,
    compress, decompress,
    create_compression_context, compress_begin, compress_update, compress_end,
    create_decompression_context, get_frame_info, decompress_update,
    get_block_size
)

from .compat import Iterable as __Iterable


class Compressor(object):
    """Iteratively compress data in lz4-framed - can be used as a context manager if writing to a file, e.g.:

        with open('myFile', 'wb') as f:
            # Context automatically finalises frame on completion, unless an exception occurs
            with Compressor(f) as c:
                try:
                    while (...):
                       c.update(moreData)
                except Lz4FramedNoDataError:
                    pass

    Alternatively, with output from relevant methods:

        c = Compressor()
        while (...):
            try:
                someOutput.append(c.update(moreData))
            except Lz4FramedNoDataError:
                pass
        # Finalise frame
        someOutput.append(c.end())
    """

    def __init__(self, fp=None, block_size_id=LZ4F_BLOCKSIZE_DEFAULT, block_mode_linked=True, checksum=False,
                 autoflush=False, level=LZ4F_COMPRESSION_MIN, block_checksum=False):
        """
        Args:
            fp: File like object (supporting write() method) to write compressed data to. If not set, data will be
                returned by the update(), flush() and end() methods.
            block_size_id (int): Compression block size identifier. One of the LZ4F_BLOCKSIZE_* constants
            block_mode_linked (bool): Whether compression blocks are linked
            checksum (bool): Whether to produce frame checksum
            autoflush (bool): Whether to return (or write to fp) compressed data on each update() call rather than
                              waiting for internal buffer to be filled. (This reduces internal buffer size.)
            level (int): Compression level. Values lower than 3 (including negative ones) use fast compression.
                         Recommended range for hc compression is between 4 and 9, with a maximum of LZ4_COMPRESSION_MAX.
            block_checksum (bool): Whether to produce checksum after each block
        """
        self.__ctx = create_compression_context()
        self.__lock = Lock()
        if fp is None:
            self.__write = None
        elif not callable(fp.write):
            raise TypeError('fp.write not callable')
        else:
            self.__write = fp.write
        self.__header = compress_begin(self.__ctx, block_size_id=block_size_id, block_mode_linked=block_mode_linked,
                                       checksum=checksum, autoflush=autoflush, level=level,
                                       block_checksum=block_checksum)

    def __enter__(self):
        if self.__write is None:
            raise ValueError('Context only usable when fp supplied')
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.end()

    def update(self, b):  # pylint: disable=method-hidden,invalid-name
        """Compress data given in b, returning compressed result either from this function or writing to fp). Note:
           sometimes output might be zero length (if being buffered by lz4).
           Raises Lz4FramedNoDataError if input is of zero length."""
        with self.__lock:
            output = compress_update(self.__ctx, b)
            if self.__write:
                self.__write(self.__header)
                self.__header = None
                self.__write(output)
                self.update = self.__updateNextWrite
                return None

            header = self.__header
            self.__header = None
            self.update = self.__updateNextReturn
            return header + output

    # post-first update methods so do not require header write & fp checks
    def __updateNextWrite(self, b):  # pylint: disable=invalid-name
        self.__write(compress_update(self.__ctx, b))

    def __updateNextReturn(self, b):  # pylint: disable=invalid-name
        return compress_update(self.__ctx, b)

    def end(self):
        """Finalise lz4 frame, outputting any remaining as return from this function or by writing to fp)"""
        with self.__lock:
            if self.__write:
                self.__write(compress_end(self.__ctx))
                return None

            return compress_end(self.__ctx)


class Decompressor(__Iterable):  # pylint: disable=super-init-not-called
    """Iteratively decompress blocks of an lz4-frame from a file-like object, e.g.:

        with open('myFile', 'rb') as f:
            try:
                for chunk in Decompressor(f):
                   decoded.append(chunk)
            except Lz4FramedNoDataError:
                # Frame incomplete - error case

    The decompressor will automatically choose a meaningful read size. Note that some
    iterator calls might return zero-length data. The iterator raises LZ4FNoDataError
    if input (from fp.read) is of zero length, before decompression finished.
    """

    def __init__(self, fp):
        """
        Args:
            fp: File like object (supporting read() method) to read compressed data from.
        """
        super(Decompressor, self).__init__()
        if fp is None:
            raise TypeError('fp')
        elif not callable(fp.read):
            raise TypeError('fp.read not callable')
        else:
            self.__read = fp.read
        self.__info = None
        self.__ctx = create_decompression_context()
        self.__lock = Lock()

    def __iter__(self):
        ctx = self.__ctx
        read = self.__read
        input_hint = 15  # enough to read largest header
        chunk_size = 32  # output chunk size, will be increased once block size known

        with self.__lock:
            output = decompress_update(ctx, read(input_hint), chunk_size)
            try:
                self.__info = info = get_frame_info(ctx)
            except Lz4FramedError as ex:
                if ex.args[1] != LZ4F_ERROR_frameHeader_incomplete:
                    # should not happen since have read 15 bytes
                    raise
            else:
                chunk_size = get_block_size(info['block_size_id'])
            input_hint = output.pop()

            # return any data as part of header read, if present
            for element in output:
                yield element

            while input_hint > 0:
                output = decompress_update(ctx, read(input_hint), chunk_size)
                input_hint = output.pop()
                for element in output:
                    yield element

    @property
    def frame_info(self):
        """See get_frame_info(). Note: This will return None if not enough data has been
           read yet to decode header (typically at least one read from iterator)."""
        return self.__info
